### PR TITLE
Implement command completion in console for openhab command

### PR DESCRIPTION
**Pull request by ccutrer**
And subcommands can implement it if they want. At the least the CommandConsoleExtension's description will now be visible in tab-completion.

I've also implemented detailed completion for openhab:items, openhab:status, openhab:send, and openhab:update.

### CODE REVIEW

File: bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/Completer.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * Implementing this interface allows a {@link ConsoleCommandExtension} to
+ * provide completions for the user as they write commands.
+ *
+ * @author Kai Kreuzer - Initial contribution
+ */
+@NonNullByDefault
+public interface Completer {
```
Comment (J-N-K) [inline bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/Completer.java:26]: I would suggest to rename this class so the conflict with `org.apache.karaf.shell.api.console.Completer` is avoided. WDYT about `ConsoleCommandCompleter`?

File: bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.karaf.shell.support.completers.StringsCompleter;
+import org.eclipse.jdt.annotation.NonNull;
+import org.openhab.core.io.console.Completer;
+import org.openhab.core.io.console.extensions.ConsoleCommandExtension;
+
+/**
+ *
+ * @author Cody Cutrer - Initial contribution
+ */
+public class CompleterWrapper implements org.apache.karaf.shell.api.console.Completer {
```
Comment (J-N-K) [inline bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java:30]: Is it possible to annotate the class with `@NonNullByDefault`? Also please add a class description in JavaDoc

File: bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/StringsCompleter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * Completer for a set of strings
+ *
+ * @author Cody Cutrer - Initial contribution
+ */
+@NonNullByDefault
+public class StringsCompleter implements Completer {
+    private final SortedSet<String> strings;
+    private final boolean caseSensitive;
+
+    public StringsCompleter() {
```
Comment (J-N-K) [inline bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/StringsCompleter.java:33]: Why do you need so many constructors? IMO the default constructor and `public StringsCompleter(final Collection<String> strings, boolean caseSensitive)` would be sufficient.

File: bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/extensions/ConsoleCommandExtension.java [hunk 4]
```diff
@@ -53,4 +55,11 @@ public interface ConsoleCommandExtension {
      * @return the help texts for this extension
      */
     List<String> getUsages();
+
+    /**
+     * @return a {@link Completer} object for this command
```
Comment (J-N-K) [inline bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/extensions/ConsoleCommandExtension.java:60]: Please add a method description

**Review by J-N-K — CHANGES_REQUESTED** [review]
In general looks good and a nice addition. Please check that you have JavaDoc for all `public` methods and add some tests.

File: bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java (hunk 2 above)
Comment (ccutrer) [reply bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java:30]: I was having trouble with that. The interface is not `@NonNullByDefault`, and I couldn't figure out the proper encantation of `@Nullable` on the overridden methods to make it compatible. And given that the only methods in this are overridden from the interface, I didn't think it was worth it.

File: bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/StringsCompleter.java (hunk 3 above)
Comment (ccutrer) [reply bundles/org.openhab.core.io.console/src/main/java/org/openhab/core/io/console/StringsCompleter.java:33]: 🤷  this class is mostly based off of (one of the several) StringsCompleter classes from karaf, and they had a bunch of constructors too. I have no problem trimming it down.

File: bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java (hunk 2 above)
Comment (J-N-K) [reply bundles/org.openhab.core.io.console.karaf/src/main/java/org/openhab/core/io/console/karaf/internal/CompleterWrapper.java:30]: Ok. Probably `@NonNullByDefault({})` would have done the trick.

**Comment by ccutrer** [comment]
I've addressed the comments so far, except for writing tests. I don't have time to do that today. Hopefully within the next handful of days.

**Review by J-N-K — APPROVED** [review]
LGTM, thanks.

**Comment by lolodomo** [comment]
This is a really cool feature, bravo.
We have console commands implemented in 14 bindings.
I am using and maintaining (or able to maintain) 7 of them. So I will extend your solution to the following bindings: hue, netatmo, linky, powermax, freebox, lgwebos, somfytahoma.

**Pull request merged**